src/main/java/org/openstreetmap/atlas/checks/validation/intersections/LineCrossingWaterBodyCheck.java

```diff
@@ -393,11 +396,21 @@ private boolean canCrossWaterBody(final AtlasItem crossingItem,
                 // and the waterbody (or its outer member components if it's a multipolygon)
                 final Set<Tuple<PolyLine, Set<Location>>> interactionsPerWaterbodyComponent = this
                         .getInteractionsPerWaterbodyComponent(waterbody, object,
-                                lineItem.asPolyLine());
+                                lineItem.asPolyLine(), flag);
+
                 // Just need to see if the intersection points are allowed in OSM; if not flag them
                 return !interactionsPerWaterbodyComponent.isEmpty()
-                        && interactionsPerWaterbodyComponent.stream().anyMatch(
-                                tuple -> !this.canCrossWaterBody(lineItem, waterbody, tuple));
+                        && interactionsPerWaterbodyComponent.stream().anyMatch(tuple ->
```

Comment:
Sorry I just noticed this- when you're removing points from the flag here, we short circuit if we run into a !validCrossWaterbody, when really we want to remove all valid points from the flag THEN return if any of interactionsPerWaterbodyComponent is invalid. So one thing to do would be move validCrossWaterbody to be an AtomicBoolean outside the lambda function, and set it to false if `this.canCrossWaterBody(lineItem, waterbody, tuple)` is ever false, then return validCrossWaterbody after the lambda function

Reply:
Addressed! (I think)